### PR TITLE
Refactor lateral lag compensation

### PR DESCRIPTION
**Pull request by haraschax**

### CODE REVIEW

File: selfdrive/controls/lib/drive_helpers.py [hunk 1]
```diff
@@ -52,3 +60,33 @@ def initialize_v_cruise(v_ego, buttonEvents, v_cruise_last):
       return v_cruise_last
 
   return int(round(clip(v_ego * CV.MS_TO_KPH, V_CRUISE_ENABLE_MIN, V_CRUISE_MAX)))
+
+
+def get_lag_adjusted_curvature(CP, v_ego, psis, curvatures, curvature_rates):
+  if not len(psis) == CONTROL_N:
```
Comment (adeebshihadeh) [inline selfdrive/controls/lib/drive_helpers.py:66]:
```suggestion
  if len(psis) != CONTROL_N:
```

File: selfdrive/controls/controlsd.py [hunk 2]
```diff
@@ -580,7 +576,7 @@ def publish_logs(self, CS, start_time, actuators, v_acc, a_acc, lac_log):
     controlsState.enabled = self.enabled
     controlsState.active = self.active
     controlsState.curvature = curvature
-    controlsState.steeringAngleDesiredDeg = angle_steers_des
+    controlsState.steeringAngleDesiredDeg = actuators.steeringAngleDeg
```
Comment (adeebshihadeh) [inline selfdrive/controls/controlsd.py:579]: deprecate the field?

Comment (HaraldSchafer) [reply selfdrive/controls/controlsd.py:579]: We don't want it?

Comment (adeebshihadeh) [reply selfdrive/controls/controlsd.py:579]: duplicating fields across packets generally isn't great. any reason to keep it?

Comment (HaraldSchafer) [reply selfdrive/controls/controlsd.py:579]: ah I see its already in carcontrol, ok!

File: selfdrive/test/process_replay/regen.py [hunk 3]
```diff
@@ -165,22 +166,25 @@ def regen_segment(lr, frs=None, outdir=FAKEDATA):
...
-  lr = LogReader(r.log_paths()[args.seg])
-  fr = FrameReader(r.camera_paths()[args.seg])
-  rpath = regen_segment(lr, {'roadCameraState': fr})
-  relr = os.path.relpath(rpath)
-
-  print("\n\n", "*"*30, "\n\n")
-  print("New route:", relr, "\n")
-  if args.upload:
+if __name__ == "__main__":
+  new_segments = []
```
Comment (adeebshihadeh) [inline selfdrive/test/process_replay/regen.py:171]: let's move this to a new script. I'd like to keep the current usage

File: selfdrive/test/process_replay/regen.py [hunk 4]
```diff
@@ -18,8 +17,8 @@
 from selfdrive.manager.process import ensure_running
 from selfdrive.manager.process_config import managed_processes
 from selfdrive.test.update_ci_routes import upload_route
-from tools.lib.route import Route
 from tools.lib.framereader import FrameReader
+from selfdrive.test.process_replay.test_processes import original_segments as segments
```
Comment (adeebshihadeh) [inline selfdrive/test/process_replay/regen.py:21]: sort alphabetically

File: selfdrive/controls/lib/drive_helpers.py [hunk 5]
```diff
@@ -1,15 +1,23 @@
 from common.numpy_fast import clip, interp
 from selfdrive.config import Conversions as CV
 from cereal import car
+from selfdrive.modeld.constants import T_IDXS
+from common.realtime import DT_MDL
```
Comment (adeebshihadeh) [inline selfdrive/controls/lib/drive_helpers.py:5]: sort alphabetically

File: selfdrive/controls/lib/drive_helpers.py [hunk 6]
```diff
@@ -52,3 +60,33 @@ def initialize_v_cruise(v_ego, buttonEvents, v_cruise_last):
...
+  next_curvature_rate = curvature_rates[0]
+
+  # MPC can plan to turn the wheel and turn back before t_delay. This means
+  # in high delay cases some corrections never even get commanded. So just use
+  # psi to calculate a simple linearization of desired curvature
+  curvature_diff_from_psi = psi / (max(v_ego, 1e-1) * delay) - current_curvature
+  next_curvature = current_curvature + 2 * curvature_diff_from_psi
+
+  desired_curvature = next_curvature
+  desired_curvature_rate = next_curvature_rate
```
Comment (adeebshihadeh) [inline selfdrive/controls/lib/drive_helpers.py:84]: seems these are just copies of next_curvature and next_curvature_rate. remove them or rename next_xxx to desired_xxx?

File: selfdrive/controls/lib/latcontrol_angle.py [hunk 7]
```diff
@@ -10,14 +11,18 @@ def reset(self):
     pass
 
   def update(self, active, CS, CP, VM, params, lat_plan):
+    desired_curvature, _ = get_lag_adjusted_curvature(CP, CS.vEgo,
```
Comment (adeebshihadeh) [inline selfdrive/controls/lib/latcontrol_angle.py:14]: seems to make more sense to calculate in controlsd and pass it in

File: selfdrive/test/process_replay/regen.py (hunk 3 above)
Comment (HaraldSchafer) [reply selfdrive/test/process_replay/regen.py]: Changed back

File: selfdrive/test/process_replay/regen.py [hunk 8]
```diff
@@ -165,22 +165,24 @@ def regen_segment(lr, frs=None, outdir=FAKEDATA):
...
-  parser.add_argument("--upload", action="store_true", help="Upload the new segment to the CI bucket")
-  parser.add_argument("route", type=str, help="The source route")
-  parser.add_argument("seg", type=int, help="Segment in source route")
-  args = parser.parse_args()
 
-  r = Route(args.route)
-  lr = LogReader(r.log_paths()[args.seg])
-  fr = FrameReader(r.camera_paths()[args.seg])
+def regen_and_save(route, sidx, upload=False):
+  lr = LogReader(f"cd:/{route.replace('|', '/')}/{sidx}/rlog.bz2")
```
Comment (adeebshihadeh) [inline selfdrive/test/process_replay/regen.py:170]: this only works in the office. can you move this to regen_all.py?

**Pull request merged**